src/net_processing.cpp

```diff
@@ -2256,6 +2297,35 @@ void PeerManagerImpl::SendBlockTransactions(CNode& pfrom, Peer& peer, const CBlo
     m_connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::BLOCKTXN, resp));
 }
 
+bool PeerManagerImpl::CheckHeadersPoW(const std::vector<CBlockHeader>& headers, const Consensus::Params& consensusParams, Peer& peer)
+{
+    // Do these headers have proof-of-work matching what's claimed?
+    if (!HasValidProofOfWork(headers, consensusParams)) {
+        Misbehaving(peer, 100, "header with invalid proof of work");
+        return false;
+    }
+
+    // Are these headers connected to each other?
+    if (!CheckHeadersAreContinuous(headers)) {
+        Misbehaving(peer, 20, "non-continuous headers sequence");
+        return false;
+    }
+    return true;
+}
+
+arith_uint256 PeerManagerImpl::GetAntiDoSWorkThreshold()
+{
+    arith_uint256 near_chaintip_work = 0;
+    LOCK(cs_main);
+    if (m_chainman.ActiveChain().Tip() != nullptr) {
+        const CBlockIndex *tip = m_chainman.ActiveChain().Tip();
+        // Use a 144 block buffer, so that we'll accept headers that fork from
+        // near our tip.
```

Comment:
5691c174c5cb2f9947e6cc2db8d32b4b4715e47d: I assume we're interested in those, because although their total work is lower, there's a chance a miner builds on top of it. In that case when we receive only that new block (header), we already have the parent?

Reply:
Yeah, so when I started writing this code I think I had no buffer at all, which turns out to cause problems when there's a small reorg (requiring extra round-trips for blocks to successfully relay).  I think this also caused zillions of tests to fail.

Upon further thought, I reasoned that for anti-DoS purposes, it's enough to have any kind of work threshold that grows with the tip, rather than is pegged at the tip's work exactly.  So I picked something big enough that we should never have a problem with headers not relaying if they're at all close to the tip, while small enough that I figured an attacker would not be able to keep up for very long unless they had a huge amount of hashpower.  The exact value here is pretty arbitrary though.